### PR TITLE
#969: Inject credentials to ISettings to pass them into PushRunner

### PR DESCRIPTION
**Pull request by hemisphera**
# PR Summary
Inject credentials passed to `Publish-PSResource` into `ISettings` object in order to pass them along to the `PushRunner` that pushes the .nupkg to the feed.
Should solve #969

## PR Context
This allows publishing PS resources to an authenticated feed.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/PublishPSResource.cs [hunk 1]
```diff
@@ -922,6 +927,34 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
 
         }
 
-        #endregion
-    }
+        private void InjectCredentialsToSettings(ISettings settings, IPackageSourceProvider sourceProvider, string source)
+        {
+          if (Credential == null) return;
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs]:
```suggestion
          if (Credential == null)
          {
               return;
          }

```

File: src/code/PublishPSResource.cs [hunk 2]
```diff
@@ -922,6 +927,34 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
 
-        #endregion
-    }
+        private void InjectCredentialsToSettings(ISettings settings, IPackageSourceProvider sourceProvider, string source)
+        {
+          if (Credential == null) return;
+          var packageSource = sourceProvider.LoadPackageSources().FirstOrDefault(s => s.Source == source);
+          if (packageSource != null)
+            if (!packageSource.IsEnabled) packageSource.IsEnabled = true;
+
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs]:
```suggestion
          if (packageSource != null)
          {
                if (!packageSource.IsEnabled)
                {
                    packageSource.IsEnabled = true;
                }
           }


```

File: src/code/PublishPSResource.cs [hunk 3]
```diff
@@ -922,6 +927,34 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+          var networkCred = Credential.GetNetworkCredential();
+          string key;
+          if (packageSource == null)
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs:953]:
```suggestion
          
          if (packageSource == null)

```

File: src/code/PublishPSResource.cs [hunk 4]
```diff
@@ -922,6 +927,34 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+          {
+            key = "_" + Guid.NewGuid().ToString().Replace("-", "");
+            settings.AddOrUpdate(
+              ConfigurationConstants.PackageSources,
+              new SourceItem(key, source));
+          }
+          else
+            key = packageSource.Name;
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs]:
```suggestion
          else
          {
              key = packageSource.Name;
          }

```

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by alerickson — APPROVED** [review]

File: src/code/PublishPSResource.cs [hunk 5]
```diff
@@ -846,11 +846,16 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
 
             var settings = NuGet.Configuration.Settings.LoadDefaultSettings(null, null, null);
             var success = false;
+
+            var sourceProvider = new PackageSourceProvider(settings);
+            if (Credential != null)
+              InjectCredentialsToSettings(settings, sourceProvider, publishLocation);
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs]:
```suggestion
            if (Credential != null)
            {
                 InjectCredentialsToSettings(settings, sourceProvider, publishLocation);
            }

```

File: src/code/PublishPSResource.cs [hunk 6]
```diff
@@ -922,6 +927,49 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+            key = "_" + Guid.NewGuid().ToString().Replace("-", "");
+            settings.AddOrUpdate(
+              ConfigurationConstants.PackageSources,
+              new SourceItem(key, source));
+          }
+          else
+          {
+              key = packageSource.Name;
+          }
+
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs:965]:
```suggestion

```

File: src/code/PublishPSResource.cs [hunk 7]
```diff
@@ -922,6 +927,49 @@ private bool PushNupkg(string outputNupkgDir, string repoName, string repoUri, o
...
+
+          settings.AddOrUpdate(
+            ConfigurationConstants.CredentialsSectionName,
+            new CredentialsItem(
+              key,
+              networkCred.UserName,
+              networkCred.Password,
+              true, ""));
```
Comment (adityapatwardhan) [inline src/code/PublishPSResource.cs]: use named parameter for `true` and string.Empty for the last parameter

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by anamnavi — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by dbrennand** [comment]
Hi all,

I'd like to test this change but based on the CHANGELOG this didn't make it into `3.0.21-beta21`? #1091

**Comment by anamnavi** [comment]
@dbrennand thanks for pointing this out. I forgot to include this in the changelog, but it should be included in the beta21 code changes. We're aiming to release later today so the changes would be included when it releases.

**Pull request merged**